CodeParser/Kernel/Shims.wl

```diff
@@ -28,37 +28,44 @@ setupShims[] := (
 setupStackShim[] := (
 
   (*
-  For versions before 12.1, we are using O(n^2) AppendTo to create the stack of top-level expressions.
+  For versions before 12.1, we implement our own stack to store top-level expressions.
 
-  Limit to some low number.
+  The push, pop and peek operations take O(1), while Normal takes O(n).
   *)
-  $TopLevelExpressionLimit = 5000;
 
   (*
   Define CreateDataStructure for earlier versions
   *)
   System`CreateDataStructure["Stack"] :=
-    Module[{stack, stackVal},
-    
+    Module[{stack, stackVal, stackDepth},
+
       stackVal = {};
 
+      stackDepth = 0;
+
       stack /: stack["Push", expr_] := (
-        AppendTo[stackVal, expr];
+        stackVal = {expr, stackVal};
+        stackDepth += 1;
         Null
       );
 
       stack /: stack["Pop"] :=
-        Module[{tmp},
-          tmp = stackVal[[-1]];
-          stackVal = Drop[stackVal, -1];
-          tmp
+        Block[{tmp},
+          If[stackDepth != 0,
+            stackDepth -= 1;
+            {tmp, stackVal} = stackVal;
+            tmp
+          ]
         ];
 
-      stack /: stack["Peek"] := stackVal[[-1]];
+      stack /: stack["Peek"] := First[stackVal];
 
-      stack /: Normal[stack] := stackVal;
+      stack /: Normal[stack] := (
+        Nest[(Sow[#[[1]]]; #[[2]])&, stackVal, stackDepth]
+        // Reap // Last // Last // Reverse
```

Comment:
I would prefer doing  `Reap[...][[2, 1]]`

I don't like the style of cascading `//`